content/blog/2022-05-26-typed-assembly-language/index.md

```diff
@@ -0,0 +1,81 @@
+
+- AST type checking is done in the traditional way with a typing context, mapping identifiers to their types. Of course, the context needs to accomodate different "levels" in the AST. Since type checking might be needed for multiple stages in the compilation, I abstracted away a generic context structure.
+
+- The struct `Context<T: Clone>` lives in `util.rs`. It has functions `bind`, `get`, `enter`, and `exeuent`. Think of it as having "levels", corresponding to the parent-child relationship in any tree structure. `bind` binds an identifier to some `T` in the current level. `get` gets the value bound to some identifier. `enter` "saves" the current level and goes down one, and `exeuent` discards the current level and restores the previous level. Each level can see the mappings from all previous levels. For example, to check `e1 + e2` one would wrap the type checking of each of `e1` and `e2` between an `enter` and `exeuent`, and see if both expressions type-checked to `int`.
+
+
+- CPS translation follows pretty closely to the Morrisett paper. Instead of just expressions we now have `Expr` and `Value`. `Value` are like expressions but they cannot be further beta-reduced. The types are almost the same, except `Arrow` type is replaced by `Cont`, which doesn't have return types: CPS replaces all returns with a call to a continuation, so every function "returns" void.
+
+
+These are the features implemented so far. We expect to fill in the missing pieces in the roadmap above.
+
+
+### Evaluation
+
+Unfortunately we wasn't able to get to assembly generation, so we can't evaluate the implementation fully. However there (fairly comprehensive) integration tests for the first few steps in the compiler, which can be found in the `test` directory.
+
+I also wrote some sample SML programs in the directory `benches`, which can be later be used to test the whole compiler.
+
+To run the tests yourself, simply go to the `compiler` directory and run `cargo t`, provided you have Rust setup.
+
+
+
+### Difficulties
+
+There are many unexpected hiccups and difficulties, which was part of the reason why this project is still incomplete so far.
+
+First, choosing the source language was difficult. I went back and forth between typed Racket and SML. The former as a descendant of LISP and Scheme is syntactically closer to System F, and it has explicit type abstraction and type application, which makes it nice. However, its type system, especially the numeric tower and stuff, is a bit too convoluted. SML is more familiar for me since I have programmed in OCaml a lot, but it depends on type inference (unification). I eventually went with SML since its type system makes sense, and to deal with unification I required explicit type annotation in the syntax. However this still leaves out implicit type abstraction and application, which is why I had to disable the polymorphic function feature for the time being.
```

Comment:
I think this is probably an important lesson to take away. The interesting part of the project didn't really have to do with the choice of source language per se—a hacked-up custom implementation of pure System F would have been fine. In compiler implementation, it's usually a good idea to "cheat" on all the inessential stuff for as long as possible so you can focus on the hard problems—here, this might have looked like giving up (at least at first) on source-code compatibility with any extant language.

Reply:
I agree. In hindsight I should done what you suggested from the start. I recently switched the grammar to an extended version of the System F from 4110, so that should make everything else following that easier.

My main motivation for compiling from industry languages was so that I get a bunch of benchmarks for free. However I overlooked the fact that "real" languages have a bunch of QOL features that are nice for programmers but add extra headache for compiler writers.